int32 merged_num
int32 source_num
